### PR TITLE
tests: allow arbitrary S3 hosts, and use minio for cassettes

### PR DESCRIPTION
**Pull request by LawnGnome**
As part of this work, I've chosen to streamline the `RECORD` environment variable down to only allowing `yes` as an option: the `passthrough`/`force` split added in #5040 doesn't really make sense to me, and it feels unnecessary in practice. Either we're using the cassettes or we're updating them.

I've also written some new documentation on integration tests. Right now it's basically just covering the `vcr` pattern and how we use it, but it's a starting point. There should be enough there for a contributor to be able to regenerate the entire set of cassettes, should they wish to.

I've tested this against both a local minio (which has been used to regenerate the cassettes) and a real S3 bucket. Both appear to work as expected. Nevertheless, I strongly believe that we should make the cassettes included in the repository be based on a local minio in future to improve the developer experience and make it easier to update these for real.

One lingering thing I'm not so sure about are the actual cassette updates themselves: minio evidently returns a bunch of response headers that we didn't previously have encoded in the cassettes. I'm happy to add them to the ignored headers list and regenerate the cassettes if we want to get rid of the noise.

### CODE REVIEW

**Comment by LawnGnome** [comment]
Thanks for the rebase, @Turbo87!

This is now ready for review.

File: docs/BACKEND.md [hunk 1]
```diff
@@ -114,4 +114,65 @@ matter what language or framework crates.io was implemented in.
...
+to the upstream service, and the cassette will be updated with the new response.
+
+In addition to setting `RECORD`, you will also need to set a number of other
+environment variables to handle any requests to S3. These are in `.env.sample`,
+but are also reproduced here for convenience:
+
+* `TEST_S3_BUCKET`: the S3 bucket used for package uploads.
+* `TEST_S3_REGION`: the S3 region used to package uploads. This may also be an
+  absolute URL (such as `http://127.0.0.1:19000`), in which that will be used as
+  the endpoint.
```
Comment (Turbo87) [inline docs/BACKEND.md:148]: does it make sense for this to be configurable? the URL ends up in the JSON files, so using something different would cause the tests to fail (see below) 🤔

File: src/config/base.rs [hunk 2]
```diff
@@ -139,3 +140,40 @@ impl Base {
...
+    match maybe_region {
+        Some(region) if region.contains("://") => {
+            let (_proto, host) = region.split_once("://").unwrap();
+            s3::Region::Host(host.to_string())
+        }
+        Some(region) if region.is_empty() => s3::Region::Default,
+        Some(region) => s3::Region::Region(region),
+        // This needs to match the region configuration that was used to generate the cassettes in
+        // `src/tests/http-data`.
+        None => s3::Region::Host(DEFAULT_TEST_REGION.to_string()),
```
Comment (Turbo87) [inline src/config/base.rs:156]: it seems a bit unexpected to me that "empty string" and "no string" behave entirely differently 🤔

File: docs/BACKEND.md (hunk 1 above)
Comment (LawnGnome) [reply docs/BACKEND.md:148]: I think so: what it does do is allow a test run against real S3 with `RECORD=yes` — while the result of that isn't something we'd want to commit, it means we can do a test run against the real service and be comfortable that we still get real responses back.

File: src/config/base.rs (hunk 2 above)
Comment (LawnGnome) [reply src/config/base.rs:156]: Yeah, that's probably too confusing on second thought. I'm going to change it so an empty string is handled identically to a missing one, and anyone who wants to use the "default" region can just say `us-east-1` explicitly.

File: crates_io_s3/lib.rs [hunk 3]
```diff
@@ -42,7 +42,7 @@ impl Region {
         Ok(Url::parse(&match self {
             Region::Host(host) => format!("{proto}://{host}/{bucket}/{path}"),
             Region::Region(region) => {
-                format!("{proto}://{bucket}.s3-{region}.amazonaws.com/{path}")
+                format!("{proto}://{bucket}.s3.{region}.amazonaws.com/{path}")
```
Comment (Turbo87) [inline crates_io_s3/lib.rs:45]: this change seems quite unrelated to the other changes here and something we might want to test independently from the HTTP recording stuff on staging/prod first.

Comment (LawnGnome) [reply crates_io_s3/lib.rs:45]: 35950ea has the full detail on why this change was made — basically, it allows `us-east-1` (and new S3 regions) to be used, since they don't have the same hyphenated region CNAMEs as some other "legacy" regions.

Do you want me to cherry pick this out into another PR so we can test it on staging?

Comment (LawnGnome) [reply crates_io_s3/lib.rs:45]: > does this change block the other changes from this PR? do we need to support the non-legacy regions to make the HTTP recordings work?

Eventually, yes — not supporting them means we don't support `us-east-1`, which is probably important and a lot of people default to — but it's not _immediately_ urgent.

So what I'll do is:

1. Back out the hyphen-to-subdomain change and move it to a different PR.
2. Document that restriction for now in `BACKEND.md` and `.env.sample`.
3. Figure out what's up with that failing backend test.
4. ...
5. ~~Profit?~~ Bug you to merge. :laughing:

File: src/tests/record.rs [hunk 4]
```diff
@@ -284,17 +287,30 @@ async fn record_http(req: Request<Body>, client: Client) -> Result<ResponseAndEx
...
         // Construct an outgoing request
-        let uri = uri.to_string().replace("http://", "https://");
-        let uri = uri.parse::<Uri>().unwrap();
         let mut req = Request::builder()
             .method(method.clone())
             .uri(uri)
             .body(body.into())
             .unwrap();
         *req.headers_mut() = headers.clone();
+        info!(?req);
```
Comment (Turbo87) [inline src/tests/record.rs:313]: was this intentionally left in? (same for the `warn!()` above)

**Review by Turbo87 — APPROVED** [review]
quick! let's merge this while CI is still green!! 😂

File: src/tests/record.rs (hunk 4 above)
Comment (LawnGnome) [reply src/tests/record.rs:313]: …fuck.

File: src/config/base.rs [hunk 5]
```diff
@@ -139,3 +140,39 @@ impl Base {
         }
     }
 }
+
+static DEFAULT_TEST_REGION: &str = "127.0.0.1:19000";
+
+fn parse_region(maybe_region: Option<String>) -> s3::Region {
```
Comment (jtgeibel) [inline src/config/base.rs:146]: This is currently only called from `Base::test()` so the unwrap below is fine. I'm sure we're not entirely consistent, but when using unwrap/expect we try to include a doc comment with a `# Panics` section so that it is clear this shouldn't be called from production code.

(I think we have a few exceptions where we can panic in startup code if the configuration is invalid, so that exception would likely apply here as well. Just that in particular, we try to never risk a panic in middleware/endpoints, even though those should just result in a status 500.)

**Review by jtgeibel — APPROVED** [review]
This is great! Just one minor comment.
